pulsar/transaction_coordinator_client.go

```diff
@@ -20,23 +20,330 @@ package pulsar
+		RequestId: proto.Uint64(requestID),
+		TcId:      proto.Uint64(t.partition),
+	}
+
+	res, err := t.tc.client.rpcClient.Request(lr.LogicalAddr, lr.PhysicalAddr, requestID,
+		pb.BaseCommand_TC_CLIENT_CONNECT_REQUEST, &cmdTCConnect)
+
+	if err != nil {
+		t.log.WithError(err).Error("Failed to connect transaction_impl coordinator " +
+			strconv.FormatUint(t.partition, 10))
+		return err
+	}
+
+	go func() {
+		select {
+		case <-t.closeCh:
+			return
+		case <-res.Cnx.WaitForClose():
+			t.connectClosedCh <- &connectionClosed{}
+		}
+	}()
+	t.conn.Store(res.Cnx)
+	t.log.Infof("Transaction handler with transaction coordinator id %d connected", t.partition)
+	return nil
+}
+
+func (t *transactionHandler) getConn() internal.Connection {
+	return t.conn.Load().(internal.Connection)
+}
+
+func (t *transactionHandler) runEventsLoop() {
+	for {
+		select {
+		case <-t.closeCh:
+			return
+		case req := <-t.requestCh:
+			switch r := req.(type) {
+			case *newTxnOp:
+				t.newTransaction(r)
+			case *addPublishPartitionOp:
+				t.addPublishPartitionToTxn(r)
+			case *addSubscriptionOp:
+				t.addSubscriptionToTxn(r)
+			case *endTxnOp:
+				t.endTxn(r)
+			}
+		case <-t.connectClosedCh:
+			t.log.Infof("Transaction handler %d will reconnect to the transaction coordinator", t.partition)
+			t.reconnectToBroker()
+		}
+	}
+}
+
+func (t *transactionHandler) reconnectToBroker() {
+	var delayReconnectTime time.Duration
+	var defaultBackoff = internal.DefaultBackoff{}
+
+	for {
+		if t.getState() == txnHandlerClosed {
+			// The handler is already closing
+			t.log.Info("transaction handler is closed, exit reconnect")
+			return
+		}
+
+		delayReconnectTime = defaultBackoff.Next()
+
+		t.log.WithFields(log.Fields{
+			"delayReconnectTime": delayReconnectTime,
+		}).Info("Reconnecting to broker")
+		time.Sleep(delayReconnectTime)
+
+		// double check
+		if t.getState() == txnHandlerClosed {
+			// Txn handler is already closing
+			t.log.Info("transaction handler is closed, exit reconnect")
+			return
+		}
+
+		err := t.grabConn()
+		if err == nil {
+			// Successfully reconnected
+			t.log.Info("Reconnected transaction handler to broker")
+			return
+		}
+		t.log.WithError(err).Error("Failed to create transaction handler at reconnect")
+		errMsg := err.Error()
+		if strings.Contains(errMsg, errMsgTopicNotFound) {
+			// when topic is deleted, we should give up reconnection.
+			t.log.Warn("Topic Not Found.")
+			break
+		}
+	}
+}
+
+func (t *transactionHandler) checkRetriableError(err error, op any) bool {
+	if err != nil && errors.Is(err, internal.ErrConnectionClosed) {
+		// We are in the EventLoop here, so we need to insert the request back to the requestCh asynchronously.
+		go func() {
+			t.requestCh <- op
+		}()
+		return true
+	}
+	return false
+}
+
+type newTxnOp struct {
+	// Request
+	timeout time.Duration
+
+	// Response
+	done  chan any
+	err   error
+	txnID *TxnID
+}
+
+func (t *transactionHandler) newTransaction(op *newTxnOp) {
```

Comment:
We should consider renaming these types of methods to distinguish the external ones.

For example:

- **internalXX**
- **retriableXX**

Reply:
They are implemented under different structs. It can be distinguished by the struct.